clang/test/SemaSYCL/forward-decl.cpp

```diff
@@ -0,0 +1,30 @@
+// RUN: %clang_cc1 -x c++ -DNOVIRTUAL -fsycl-is-device -std=c++11 -fsyntax-only -verify -pedantic %s
+// RUN: %clang_cc1 -x c++ -DVIRTUAL -fsycl-is-device -std=c++11 -fsyntax-only -verify -pedantic %s
+
+template <typename name, typename Func>
+__attribute__((sycl_kernel)) void kernel_single_task(Func kernelFunc) {
+  kernelFunc();
+}
+
+int main() {
+#ifdef NOVIRTUAL
+  kernel_single_task<class kernel_function_1>([]() {
+      // expected-no-diagnostics
+      class Foo *F;
+  });
+#elif VIRTUAL
+  kernel_single_task<class kernel_function_2>([]() {
+      // expected-error@+2{{No class with a vtable can be used in a SYCL kernel or any code included in the kernel}}
+      // expected-note@+1{{used here}}
+      class Boo {
+      public:
+        virtual int getBoo() { return 42; }
+      };
+  });
+
+  kernel_single_task<class kernel_function_3>([]() {
+      class Boo *B;
```

Comment:
Looks like I didn't get something but this test case is same as on line 13, since you declared local classes. This `Boo` is not same class which you declared on line 19.